Add Prometheus support to Prebid Cache

guscarreon opened the pull request:
With SIN3 we're moving to using Prometheus for metrics and eventually move all other data centers to using Prometheus as well. And so, we want to add support to send metrics to Prometheus for Prebid Cache.

hhhjort commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,172 @@
+}
+type PrometheusMetricsEntry struct {
+	Duration   *prometheus.HistogramVec
+	Errors     prometheus.Counter
```
Why three counters rather than a vector with labels?

guscarreon replied:
I rewrote the `PrometheusMetrics` object in a way I believe better adapts to the Prometheus metric paradigms. Please let me know what you think of this new approach. I can always roll back.

hhhjort commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,172 @@
+
+type PrometheusMetricsEntryByFormat struct {
+	Duration       metrics.Histogram
+	Errors         prometheus.Counter
```
Again any reason for not using vectors?

guscarreon replied:
Everything was refactored to make better use of the `CounterVec` objects. Please let me know what you think of this new approach. I can always roll back.

hhhjort commented on config/config.go:
```diff
@@ -136,19 +136,27 @@ const (
+		case MetricsInflux:
+			cfg.Influx.validateAndLog()
+		default:
+			log.Fatalf(`invalid config.metrics.type: %s. It must be "none" or "influx"`, metricsType)
```
Message needs to be updated.

guscarreon replied:
I'll correct it soon

hhhjort commented on config/config.go:
```diff
@@ -167,15 +175,21 @@ type Influx struct {
+func (cfg *PrometheusMetrics) validateAndLog() {
+	log.Infof("config.metrics.prometheus.namespace: %s", cfg.Namespace)
+	log.Infof("config.metrics.prometheus.subsystem: %s", cfg.Subsystem)
 	// This intentionally skips username and password for security reasons.
```
Does prometheus even have a user name and password option?

guscarreon replied:
I'll correct this soon

guscarreon replied:
Corrected

SyntaxNode commented on metrics/core.go:
```diff
@@ -117,3 +94,56 @@ func CreateMetrics() *Metrics {
+	registry := prometheus.NewRegistry()
+	promMetrics := &PrometheusMetrics{
+		Registry: registry,
+		RequestDurationMetrics: newHistogramVector(cfg, registry,
```
If `method` here is PUT vs GET, I would recommend to separate these out into their own metrics. Read and write operations are inherently separate and I don't see value in sharing the same metric name. I also think we want to separate front end (calls into Prebid Cache) vs backend requests (calls out to the data store). 

What are the possible values for the `result` label? We need to be careful since each one will use 13 counters (10 for the buckets, 1 for the infinity bucket, 2 for the counter and sum counters).

guscarreon replied:
corrected

SyntaxNode commented on metrics/core.go:
```diff
@@ -117,3 +94,56 @@ func CreateMetrics() *Metrics {
+			[]string{"method", "result"},
+			cacheWriteTimeBuckts,
+		),
+		MethodToEndpointMetrics: newCounterVecWithLabels(cfg, registry,
```
Similarly, consider breaking out the `method` dimension and using this to count the number of requests and labeling them with a status such as bad and error requests.

guscarreon replied:
corrected

SyntaxNode commented on metrics/core.go:
```diff
@@ -117,3 +94,56 @@ func CreateMetrics() *Metrics {
+			[]string{"method"},
+			cacheWriteTimeBuckts,
+		),
+		ConnectionErrorMetrics: newCounterVecWithLabels(cfg, registry,
```
In Prebid Server, we count connections opened and connections closed separately, which can be combined on the server to produce the current open connection counter. Consider using that approach here, which eliminates the need for the `ActiveConnections` gauge.

guscarreon replied:
corrected

SyntaxNode commented on metrics/core.go:
```diff
@@ -117,3 +94,56 @@ func CreateMetrics() *Metrics {
+			"connections.active_incoming",
+			"How many connections are currenctly opened",
+		),
+		ExtraTTLSeconds: prometheus.NewHistogram(prometheus.HistogramOpts{
```
Why do we track this metric?

guscarreon replied:
InfluxDB metrics were already tracking this field. Do you want us to reconsider the use of this metric in a future PR?

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,165 @@
+	"github.com/vrischmann/go-metrics-influxdb"
+)
+
+type MetricsEntry struct {
```
Another way of viewing how this MetricsEntry might look for Prometheus is with...
 - Duration Histogram (non-vectorized). That would use 13 counters if you specify 10 buckets.
 - Request Counter (vectorized) with a status label of values 'ok', 'badRequest', and 'error'. I think that would use 3 counters.

guscarreon replied:
Ok. Just to clarify:
```
 35 type PrometheusMetricsEntry struct {
 36   - Duration   *prometheus.HistogramVec
 37   - Errors     prometheus.Counter
 38   - BadRequest prometheus.Counter
 39   - Request    prometheus.Counter
 40
 41   + Duration   prometheus.Histogram  //Non vector
 42   + RequestStatus *prometheus.CounterVec  // CounterVec "status": "ok", "error", or "bad_request"
 43   + /*initialized like this:
 44   + metrics.requests = newCounter(cfg, metrics.Registry,
 45   +     "requests",
 46   +     "Count of total requests to Prebid Server labeled by status.",
 47   +     []string{"status"})
 48   + */
 49 }
metrics/prometheus.go [+]
```
Right?

SyntaxNode replied:
Yup. Maybe name it simply `Requests`.

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,165 @@
+	Request    metrics.Meter
+}
+
+type MetricsEntryByFormat struct {
```
Another way of viewing how this MetricsEntryByFormat might look for Prometheus is with...
 - Request Duration Histogram (non-vectorized).  That would use 13 counters if you specify 10 buckets.
 - Request Counter (vectorized) with a type label of values 'json' and 'xml' and a status label of values 'ok', 'badRequest',  'error', and 'invalid'. I think that would use 8 counters.
 - Request Ttl Counter with a status of label of 'present' or 'missing'. Do we really even need this though?
 - Request Duration Histogram (non-vectorized).  That would use 13 counters if you specify 10 buckets.. or perhaps with a type label for json vs xml which brings us to 26 counters.

guscarreon replied:
Just to make sure I understood correctly. You mean like this?
```
 27 type PrometheusMetrics struct {
 28     Registry        *prometheus.Registry
 29     Puts            *PrometheusMetricsEntry
 30     Gets            *PrometheusMetricsEntry
 31     PutsBackend     *PrometheusMetricsEntryByFormat
 32     GetsBackend     *PrometheusMetricsEntry
 33     Connections     *PrometheusConnectionMetrics
 34     ExtraTTLSeconds *prometheus.HistogramVec
 35 }
 36 +-- 14 lines: type PrometheusMetricsEntry struct {---------------------------------------------------------------------------------------------------
 50
 51 type PrometheusMetricsEntryByFormat struct {
 52     Duration       metrics.Histogram  //Non vector
 53   - Errors         prometheus.Counter // CounterVec "type": "json", "status": "error"
 54   - BadRequest     prometheus.Counter // CounterVec "type": "json", "status": "bad_request"
 55   - JsonRequest    prometheus.Counter // CounterVec "type": "json", "status": "ok"
 56   - XmlRequest     prometheus.Counter // CounterVec "type": "xml",  "status": "ok"
 57   - DefinesTTL     prometheus.Counter // CounterVec "type": "json", "status": "ok"
 58   - InvalidRequest prometheus.Counter // CounterVec "type": "json", "status": "invalid"
 59   + PutBackendRequests *prometheus.CounterVec // CounterVec "format": "json" or  "xml","status": "ok", "error", or "bad_request","definesTimeToLive">
 60   + /*initialized like this:
 61   +   promehteusMetrics.metricsEntryByFormat.PutBackendRequests = newCounter(cfg, metrics.Registry,
 62   +     "put_backend_requests",
 63   +     "Count of total requests to Prebid Cache labeled by format, status and whether or not it comes with TTL",
 64   +     []string{"format","status","definesTimeToLive"})
 65   + */
 66     RequestLength  metrics.Histogram  //Non vector
 67 }
metrics/prometheus.go [+]
```

guscarreon replied:
Or like:
```
 70 type PrometheusMetricsEntryByFormat struct {
 71     Duration       metrics.Histogram  //Non vector
 72   - Errors         prometheus.Counter // CounterVec "type": "json", "status": "error"
 73   - BadRequest     prometheus.Counter // CounterVec "type": "json", "status": "bad_request"
 74   - JsonRequest    prometheus.Counter // CounterVec "type": "json", "status": "ok"
 75   - XmlRequest     prometheus.Counter // CounterVec "type": "xml",  "status": "ok"
 76   - DefinesTTL     prometheus.Counter // CounterVec "type": "json", "status": "ok"
 77   - InvalidRequest prometheus.Counter // CounterVec "type": "json", "status": "invalid"
 78   + backendJsonRequests *prometheus.CounterVec // CounterVec "status": "ok", "error", or "bad_request","definesTimeToLive": "TTL_present", or "TTL_m>
 79   + /*initialized like this:
 80   +   promehteusMetrics.metricsEntryByFormat.backendJsonRequests = newCounter(cfg, metrics.Registry,
 81   +     "put_backend_requests",
 82   +     "Count of total JSON requests to Prebid Cache labeled by status and whether or not it comes with TTL",
 83   +     []string{"status","definesTimeToLive"})
 84   + */
 85   + backendXmlRequests *prometheus.CounterVec // CounterVec "status": "ok", "error", or "bad_request","definesTimeToLive": "TTL_present", or "TTL_mi>
 86   + /*initialized like this:
 87   +   promehteusMetrics.metricsEntryByFormat.backendXmlRequests = newCounter(cfg, metrics.Registry,
 88   +     "put_backend_requests",
 89   +     "Count of total XML requests to Prebid Cache labeled by status and whether or not it comes with TTL",
 90   +     []string{"status","definesTimeToLive"})
 91   + */
 92     RequestLength  metrics.Histogram  //Non vector
 93 }
metrics/prometheus.go [+]
```

guscarreon replied:
> Request Ttl Counter with a status of label of 'present' or 'missing'. Do we really even need this though?
>
That field is defined on the Influx metrics side. I supposed I should keep the Influx metrics on the Prometheus side as well. Do you want me to get rid of it?

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,165 @@
+	RequestLength  metrics.Histogram
+}
+
+type ConnectionMetrics struct {
```
For Prometheus, I highly recommend to record two counters, one for opened and one for closed metrics. This is what we do in Prebid Server.

guscarreon replied:
Got it, do you also want to add connections closed, as we do in prebid-server?
```
 54 type PrometheusConnectionMetrics struct {
 55     - ActiveConnections      prometheus.Counter
 56     - ConnectionCloseErrors  prometheus.Counter
 57     - ConnectionAcceptErrors prometheus.Counter
 58
 59     + ConnectionsOpened   prometheus.Counter
 60     + ConnectionsErrors   *prometheus.CounterVec // the "Connection_error" label will hold the values "accept" or "close"
 61 }
metrics/prometheus.go
```

SyntaxNode replied:
Yeah. You really just need connections opened and closed. Active can be inferred.

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,165 @@
+
+type InfluxMetrics struct {
+	Registry        metrics.Registry
+	Puts            *MetricsEntry
```
I still like separating incoming puts, incoming gets, db puts, and db gets into their own metrics for Prometheus.

guscarreon replied:
Ok, rolling back

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,165 @@
+
+func (m InfluxMetrics) Increment(metricName string, start *time.Time, value string) {
+	switch metricName {
+	case "puts.current_url.request_duration":
```
Why do you prefer to use string identifiers here? Consider breaking these out into methods and let the metric provider use their discretion. I think that approach works well in Prebid Server, with exception of the very large interface contract.

This gives more freedom for Prometheus to record things differently from Influx under the covers.

guscarreon replied:
Done

hhhjort commented on Gopkg.lock:
```diff
@@ -2,12 +2,15 @@
 
```
We should probably merge the other PR that changes Gopkg.lock before this one.

SyntaxNode replied:
Is that taken care of now? Not sure which story you were thinking of.

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,208 @@
+	if cfg.Influx.Host != "" {
+		logrus.Infof("Metrics will be exported to Influx with host=%s, db=%s, username=%s", cfg.Influx.Host, cfg.Influx.Database, cfg.Influx.Username)
+		influxdb.InfluxDB(
+			m.Registry,          // metrics registry
```
Consider pulling up the interval to a variable and removing the comments. The others all seem self explanatory.

guscarreon replied:
Done

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,208 @@
+func (m *InfluxMetrics) RecordPutRequest(status string, duration *time.Time) {
+	if status != "" {
+		switch status {
+		case "error":
```
It might be best to move these "magic numbers" to constants as is done in Prebid Server.

guscarreon replied:
Done

SyntaxNode commented on metrics/influx.go:
```diff
@@ -0,0 +1,208 @@
+			m.Puts.Request.Mark(1)
+		}
+	} else if duration != nil {
+		m.Puts.Duration.UpdateSince(*duration)
```
Why is this an `else if`? Is it not valid for both a status and a duration to be provided?

guscarreon replied:
This function gets called with one purpose or other, whether to update duration or status but not both (so far). I can change it for them to not be mutually exclusive and it wouldn't change anything in our current code. Do you prefer I remove the `else` clause?

SyntaxNode commented on metrics/metricstest/metrics_mock.go:
```diff
@@ -0,0 +1,129 @@
+)
+
+/*Define Mock metrics        */
+var HT1 map[string]float64
```
What is `HT1` vs `HT2`?

guscarreon replied:
Hash table 1 and hash table 2. Let me change it to `MockHistograms` and `MockCounters` respectively.

SyntaxNode commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,261 @@
+ *	Init functions
+ **************************************************/
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	cacheWriteTimeBuckts := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
```
Typo `Buckts` -> `Buckets`

guscarreon replied:
Changed

SyntaxNode commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,261 @@
+
+func (m *PrometheusMetrics) RecordConnectionMetrics(label string) {
+	if label == "add" {
+		m.Connections.ConnectionsOpened.Inc() //change this for  Gauge if you have time
```
No need to change. I think a counter makes sense here.

guscarreon replied:
Since the number of connections increases and decreases, a Gauge was implemented. I just forgot to remove the comments.

SyntaxNode commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,261 @@
+/**************************************************
+ *	DEPECRATED Functions to record metrics
+ **************************************************/
+func (m PrometheusMetrics) Export(cfg config.Metrics) {
```
Why is this deprecated?

guscarreon replied:
I thought it'd get deprecated but it didn't. Removed the `DEPRECATED` tag

SyntaxNode commented on metrics/prometheus.go:
```diff
@@ -0,0 +1,261 @@
+ *	NEW Functions to record metrics
+ **************************************************/
+func (m *PrometheusMetrics) RecordPutRequest(status string, duration *time.Time) {
+	incCounterInVector(m.Puts.RequestStatus, "status", status, []string{"add", "error", "bad_request"})
```
I don't understand what the slice at the end is doing.

guscarreon replied:
Because `PrometheusMetrics`, `PrometheusRequestStatusMetric`, `PrometheusRequestStatusMetricByFormat`, and `PrometheusConnectionMetric` are all objects with `prometheus.CounterVec` fields, a single function to increment their fields was created. Therefore, if the `m.Connections.ConnectionsErrors` `counterVec` has the labels: "accept_error" and "close_error", we call `incCounterInVector` like:
```
incCounterInVector(m.Connections.ConnectionsErrors, "connection_error", label, []string{"accept_error", "close_error"})
```
`m.PutsBackend.PutBackendRequests` on the other hand, has more than two labels, so this counter gets incremented like:
```
incCounterInVector(m.PutsBackend.PutBackendRequests, "label", label, []string{"add", "json", "xml", "invalid_format", "defines_ttl", "error"})
```
As a matter of fact a map[string]bool makes a lot more sense. I'll modify this.

guscarreon commented on backends/aerospike.go:
```diff
@@ -7,19 +7,19 @@ import (
-	ttlHistogram metrics.Histogram
+	cfg           config.Aerospike
+	client        *as.Client
+	metricEngines *metrics.Metrics
```
Done

guscarreon commented on backends/decorators/metrics.go:
```diff
@@ -10,49 +10,47 @@ import (
 
 func (b *backendWithMetrics) Get(ctx context.Context, key string) (string, error) {
-	b.gets.Request.Mark(1)
+	b.metricsEngines.RecGetBackendRequest("add", nil)
```
It'll add '1' to the influx or prometheus counter that keeps track of the number of backend requests. In terms of Influx, this means we add one to the influx metrics like: `InfluxMetrics.GetsBackend.Request.Mark(1)` and/or the prometheus metrics like: `PrometheusMetrics.GetsBackend.RequestStatus.With(prometheus.Labels{"status":"add"}).Inc()`

guscarreon replied:
I can change the keyword name from "add" to "increment", "mark", or something else

SyntaxNode replied:
I'm inferring that "add" = request begin, "" = request complete successfully, and "error" = request completed with error. If so, consider splitting the methods to `RecordGetBackendRequestBegin()` and `RecordGetBackendRequestFinish(err)`.

guscarreon replied:
In this part of the code I just went with the logic that was already there for Influx and function `RecGetBackendRequest(status string, duration *time.Time)` is meant to record metrics, on whatever engine we use, that relate to the "GET" requests on the backend. This function was meant to be executed like:

- `RecGetBackendRequest("add", nil)` - Will increase the counter of the get backend requests whether this request will be successful or not. It's meant to execute `b.gets.Request.Mark(1)` on influx, and `m.GetsBackend.RequestStatus.With(prometheus.Labels{label: {"add"})` on the Prometheus engine.
- `RecGetBackendRequest("", &start)` - This was a successful request for which we have an execution duration in the `&start` parameter. It's meant to execute `b.gets.Duration.UpdateSince(start)` on influx, and `m.GetsBackend.Duration.Observe(time.Since(*duration).Seconds())` on the Prometheus engine.
- `RecGetBackendRequest("error", nil)` - This was an unsuccessful request for which we have an execution duration but we don't record it, we just add to the vector tracking the number of requests that came with an error. It's meant to execute `b.gets.Errors.Mark(1)` on influx, and `m.GetsBackend.RequestStatus.With(prometheus.Labels{label: {"error"})` on the Prometheus engine.

I realize that this can be confusing. In reference to your other comment, I'll create more functions so we separate these three functionalities, and will come up with names that are more intuitive

guscarreon commented on config/config.go:
```diff
@@ -136,37 +136,75 @@ const (
 type Metrics struct {
-	Type   MetricsType `mapstructure:"type"`
-	Influx Influx      `mapstructure:"influx"`
+	Influx     Influx            `mapstructure:"influx"`
```
Done

guscarreon commented on backends/aerospike.go:
```diff
@@ -46,7 +46,7 @@ func (a *Aerospike) Get(ctx context.Context, key string) (string, error) {
 		return "", errors.New("client.Get returned a nil record. Is aerospike configured properly?")
 	}
-	a.ttlHistogram.Update(int64(rec.Expiration))
+	a.metricEngines.RecExtraTTLSeconds(float64(rec.Expiration))
```
This `ExtraTTLSeconds` metric is a histogram that records the the ttl in seconds found in aerospike for a given key value of a backend request.

guscarreon commented on config/config.go:
```diff
@@ -136,37 +136,75 @@ const (
+		metricsEnabled = cfg.Influx.validateAndLogMetricsData()
+	}
+	if cfg.Prometheus.Enabled {
+		metricsEnabled = cfg.Prometheus.validateAndLogMetricsData()
```
You are totally right. This was a bug. Corrected.

guscarreon commented on endpoints/routing/handler.go:
```diff
@@ -19,8 +19,8 @@ func NewHandler(cfg config.Configuration, dataStore backends.Backend, appMetrics
 	router.GET("/status", endpoints.Status) // Determines whether the server is ready for more traffic.
-	router.POST("/cache", decorators.MonitorHttp(endpoints.NewPutHandler(dataStore, cfg.RequestLimits.MaxNumValues, cfg.RequestLimits.AllowSettingKeys), appMetrics.Puts))
-	router.GET("/cache", decorators.MonitorHttp(endpoints.NewGetHandler(dataStore, cfg.RequestLimits.AllowSettingKeys), appMetrics.Gets))
+	router.POST("/cache", decorators.MonitorHttp(endpoints.NewPutHandler(dataStore, cfg.RequestLimits.MaxNumValues, cfg.RequestLimits.AllowSettingKeys), appMetrics, "puts"))
```
Done

guscarreon commented on endpoints/put.go:
```diff
@@ -87,7 +87,12 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
 
 			logrus.Debugf("Storing value: %s", toCache)
-			resps.Responses[i].UUID = uuid.NewV4().String()
+			u2, err := uuid.NewV4()
```
Done

guscarreon commented on metrics/influx.go:
```diff
@@ -0,0 +1,208 @@
+/**************************************************
+ *	Object definition
+ **************************************************/
+type InfluxMetrics struct {
```
Done

hhhjort commented on config/config.go:
```diff
@@ -136,37 +136,75 @@ const (
-	MetricsNone   MetricsType = "none"
-	MetricsInflux MetricsType = "influx"
-)
+type MetricsConfig interface {
```
Do you use this interface definition anywhere?

guscarreon replied:
The `MetricsConfig` is only used in this file to make `cfg.Influx` and `cfg.Prometheus` implement this interface so they have a `validateAndLogMetricsData()` function to validate we had enough info in the configuration files (or environment variables) to init the prometheus or influx metrics
```
143 func (cfg *Metrics) validateAndLog() {
144     var metricsEnabled bool = false
145     if cfg.Influx.Enabled {
146         metricsEnabled = cfg.Influx.validateAndLogMetricsData()
147     }
148     if cfg.Prometheus.Enabled {
149         metricsEnabled = cfg.Prometheus.validateAndLogMetricsData() || metricsEnabled
150     }
151     if !metricsEnabled {
152         log.Fatalf(`No metrics correctly specified in configuration file`)
153     }
154 }
155
156 type MetricsConfig interface {
157     validateAndLogMetricsData() bool
158 }
159
160 +--  7 lines: type InfluxMetrics struct {-------------------------------------------------------------------------------
167
168 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
169     if influxMetricsConfig.Host != "" {
170         influxMetricsConfig.Enabled = false
171         log.Fatalf(`Despite being enabled, influx metrics came with no host info: config.metrics.influx.host = "".`)
172     }
173     if influxMetricsConfig.Database != "" {
174         influxMetricsConfig.Enabled = false
175         log.Fatalf(`Despite being enabled, influx metrics came with no database info: config.metrics.influx.database = "".`)
176     }
177     if influxMetricsConfig.Enabled {
178         log.Infof("config.metrics.influx.host: %s", influxMetricsConfig.Host)
179         log.Infof("config.metrics.influx.database: %s", influxMetricsConfig.Database)
180     }
181     return influxMetricsConfig.Enabled
182 }
183
184 +--  7 lines: type PrometheusMetrics struct {---------------------------------------------------------------------------
191
192 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
193     if promMetricsConfig.Port == 0 {
194         promMetricsConfig.Enabled = false
195         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty port number: config.metrics.prometheus.port = 0`)
196     }
197     if promMetricsConfig.Namespace != "" {
198         promMetricsConfig.Enabled = false
199         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty name space: config.metrics.prometheus.namespace = \"\".`)
200     }
201     if promMetricsConfig.Subsystem != "" {
202         promMetricsConfig.Enabled = false
203         log.Fatalf(`Despite being enabled, prometheus metrics came with an empty subsystem value: config.metrics.prometheus.subsystem = \"\".`)
204     }
205     if promMetricsConfig.Enabled {
206         log.Infof("config.metrics.prometheus.namespace: %s", promMetricsConfig.Namespace)
207         log.Infof("config.metrics.prometheus.subsystem: %s", promMetricsConfig.Subsystem)
208     }
209     return promMetricsConfig.Enabled
210 }
config/config.go
```

hhhjort replied:
I did a search on the PR and I am not seeing it referenced anywhere. Can you point me to where you are using it?

guscarreon replied:
Sure. Objects `InfluxMetrics` and `PrometheusMetrics` are meant to implement this interface. Do you think an `interface` is too much? Want me to go with methods instead? The objective of it being an interface was to keep consistensy with more than one metrics engine.
```
168 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool
  .
  .
  .
192 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool
config/config.go
```

Get used:
```
143 func (cfg *Metrics) validateAndLog() {
144     var metricsEnabled bool = false
145     if cfg.Influx.Enabled {
146  >>     metricsEnabled = cfg.Influx.validateAndLogMetricsData()
147     }
148     if cfg.Prometheus.Enabled {
149  >>     metricsEnabled = cfg.Prometheus.validateAndLogMetricsData() || metricsEnabled
150     }
151     if !metricsEnabled {
152         log.Fatalf(`No metrics correctly specified in configuration file`)
153     }
154 }
config/config.go
```

hhhjort replied:
What I mean do you every have a variable that is of type `MetricsConfig`? The two engines do implement the method, but that is true if you define the interface or not. Having a definition for `MetricsConfig` does not guarantee that any particular object implements it.

guscarreon replied:
I meant `Influx InfluxMetrics` and `Prometheus PrometheusMetrics` to implement the `MetricsConfig` interface when defining their implementations of `validateAndLogMetricsData()`. Do you want me to get rid of the interface and just leave those  `InfluxMetrics.validateAndLogMetricsData()` and `PrometheusMetrics.validateAndLogMetricsData()` functions as methods?
```
138 type Metrics struct {
139     Influx     InfluxMetrics     `mapstructure:"influx"`
140     Prometheus PrometheusMetrics `mapstructure:"prometheus"`
141 }
142
143 +--  8 lines: func (cfg *Metrics) validateAndLog() {------------------------------
151
152 type MetricsConfig interface {
153     validateAndLogMetricsData() bool
154 }
155
156 type InfluxMetrics struct {
157     Host     string `mapstructure:"host"`
158 +--  3 lines: Database string `mapstructure:"database"`----------------------------
161     Enabled  bool   `mapstructure:"enabled"`
162 }
163
164 func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
165     if influxMetricsConfig.Host != "" {
166 +-- 11 lines: influxMetricsConfig.Enabled = false----------------------------------
177     return influxMetricsConfig.Enabled
178 }
179
180 type PrometheusMetrics struct {
181     Port             int    `mapstructure:"port"`
182 +--  3 lines: Namespace        string `mapstructure:"namespace"`-------------------
185     Enabled          bool   `mapstructure:"enabled"`
186 }
187
188 func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
189     if promMetricsConfig.Port == 0 {
190 +-- 15 lines: promMetricsConfig.Enabled = false------------------------------------
205     return promMetricsConfig.Enabled
206 }
config/config.go
```

hhhjort commented on endpoints/put.go:
```diff
@@ -87,7 +87,12 @@ func NewPutHandler(backend backends.Backend, maxNumValues int, allowKeys bool) f
 			logrus.Debugf("Storing value: %s", toCache)
-			resps.Responses[i].UUID = uuid.NewV4().String()
+			var uuidGenError error
+			resps.Responses[i].UUID, uuidGenError = utils.GenerateUUIDString()
```
You could save a line by using `:=` here I believe. I think you just need one undefined variable in the group for it to be valid go. But it is also fine the way it is.

delete line 90 and 91 is:
`resps.Responses[i].UUID, uuidGenError := utils.GenerateUUIDString()`

guscarreon replied:
I made the proposed change and I'm getting:
```
endpoints/put.go|92 col 4| : expected identifier on left side of :=
```

hhhjort replied:
That is odd, perhaps the UUID being burried in a structure is confusing things.

guscarreon replied:
@hhhjort are you ok with leaving these lines as they are? Do you want me to try something else?

hhhjort replied:
Should be fine then

hhhjort commented:
Please remove metrics/influx/.influx.go.swp from the PR. :)

hhhjort commented on metrics/influx/influx.go:
```diff
@@ -0,0 +1,223 @@
+/**************************************************
+ * Constants and global variables
+ **************************************************/
+var TenSeconds time.Duration = time.Second * 10
```
Is there a reason for these to be variables rather than constants?

guscarreon replied:
No, lol. Changed them to constants

SyntaxNode commented on backends/decorators/metrics.go:
```diff
@@ -10,49 +10,47 @@ import (
 func (b *backendWithMetrics) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
 	if strings.HasPrefix(value, backends.XML_PREFIX) {
-		b.puts.XmlRequest.Mark(1)
+		b.metricsEngines.RecPutBackendRequest("xml", nil, 0)
```
These group of 4 calls to `RecPutBackendRequest` all have nil and 0 as the second and third parameters. Consider refactoring similar to the comment above.

guscarreon replied:
done

SyntaxNode commented on config/config.go:
```diff
@@ -136,37 +136,75 @@ const (
-	default:
-		log.Fatalf(`invalid config.metrics.type: %s. It must be "none" or "influx"`, cfg.Type)
+	var metricsEnabled bool = false
+	if cfg.Influx.Enabled {
```
Would this be simpler?

```
influxEnabled := cfg.Influx.Enabled && cfg.Influx.validateAndLogMetricsData()
prometheusEnabled := cfg.Prometheus.Enabled && cfg.Prometheus.validateAndLogMetricsData()

if !influxEnabled && !prometheusEnabled {
	log.Fatalf(`No metrics correctly specified in configuration file`)
}
```

guscarreon replied:
Much simpler. Changed

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,74 @@
-	DefinesTTL     metrics.Meter
-	InvalidRequest metrics.Meter
-	RequestLength  metrics.Histogram
+/* Object to access metric engines     */
```
Please use standard Go convention for comments:

```
Metrics represents zero or more MetricEngines.
```

guscarreon replied:
Done

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,74 @@
-	metrics.RegisterRuntimeMemStats(m.Registry)
+/* Interface definition                */
+type CacheMetrics interface {
+	RecordPutRequest(status string, duration *time.Time)
```
You use `Record` here and `Rec` in other methods. Consider using `Record` everywhere for consistency.

guscarreon replied:
`RecPutRequest` is a method of the `Metrics` object which will execute the `RecordPutRequest()` implementation of the `CacheMetrics` interface objects. If we name both the `Metrics` object method, and the `CacheMetrics` interface function the same (`RecordPutRequest()`) don't we risk it'll get confusing. (Even more confusing)

SyntaxNode commented on metrics/influx/influx.go:
```diff
@@ -0,0 +1,223 @@
+	"github.com/vrischmann/go-metrics-influxdb"
+)
+
+/**************************************************
```
Consider removing the 'C' style comment blocks.

guscarreon replied:
Done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+type PrometheusRequestStatusMetric struct {
+	Duration      prometheus.Histogram   //Non vector
```
The comments in this file are useful in development, but consider removing before merging.

guscarreon replied:
Done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,272 @@
+
+type PrometheusRequestStatusMetric struct {
+	Duration      prometheus.Histogram   //Non vector
+	RequestStatus *prometheus.CounterVec // CounterVec "status": "add", "error", or "bad_request"
```
Does 'add' indicate success? I find that status string a bit confusing.

guscarreon replied:
Label string has been modified

SyntaxNode commented on backends/azure_table_test.go:
```diff
@@ -1,16 +1,19 @@
+		t.Errorf("Error generating version 4 UUID")
+	}
 
+	expected := id[0:4]
```
It looks like we only ever use the first 4 characters, consider including the shortening in `GenerateUUIDString` and rename it to `GenerateRandomID`.

guscarreon replied:
I don't think we should make `GenerateUUIDString()` to only return a 4 character string, `backends/aerospike.go`'s `Put` function uses the entire string. Did I misunderstand something? Do you mean create a function wrapper for `GenerateUUIDString()` only in `backends/azure_table_test.go` and make it return the first 4 characters only?
```
 func (a *Aerospike) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
     asKey, err := as.NewKey(a.cfg.Namespace, setName, key)  // <-- gets passed the entire key
     if err != nil {
         return err
     }
     if ttlSeconds == 0 {
         ttlSeconds = a.cfg.DefaultTTL
     }
     bins := as.BinMap{
         binValue: value,
     }
     err = a.client.Put(&as.WritePolicy{
         Expiration: uint32(ttlSeconds),
     }, asKey, bins)
     if err != nil {
         return err
     }
     return nil
 }
 backends/aerospike.go
```

guscarreon replied:
Done. `GenerateUUIDString` has been renamed to `GenerateRandomID`

hhhjort commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,272 @@
+/**************************************************
+ * Constants and global variables
+ **************************************************/
+var TenSeconds time.Duration = time.Second * 10
```
Since the structure of the metrics engine requires both Influx and prometheus to use the same constant strings, wouldn't this best be extracted to a singular `metrics_constants`, `metrics_util`, or some other shared package so you can be sure they stay in sync? This is what PBS does, it has two global metrics packages, the main one that will import all the metrics engines, and a second one for shared resources that all the metrics engines can import.

And again, variables instead of constants?

guscarreon replied:
No need to export them to a separate file becauyse prometheus and influx are now using different constants. Refactored and changed them to constants.

hhhjort replied:
No, we should be good here now.

guscarreon commented:
> Please remove metrics/influx/.influx.go.swp and metrics/prometheus/.prometheus.go.swp from the PR. :)

Sorry, done

guscarreon commented on backends/aerospike.go:
```diff
@@ -7,19 +7,19 @@ import (
 }
 
-func NewAerospikeBackend(cfg config.Aerospike, ttlHistogram metrics.Histogram) *Aerospike {
+func NewAerospikeBackend(cfg config.Aerospike, metricEngines *metrics.Metrics) *Aerospike {
```
I thought it was more descriptive. Do you want us to go for `metrics *metrics.Metrics` instead?

guscarreon replied:
I made the change

guscarreon commented on backends/decorators/metrics.go:
```diff
@@ -10,49 +10,47 @@ import (
-	puts     *metrics.MetricsEntryByFormat
-	gets     *metrics.MetricsEntry
+	delegate       backends.Backend
+	metricsEngines *metrics.Metrics
```
Done

SyntaxNode commented on endpoints/decorators/metrics.go:
```diff
@@ -28,9 +28,9 @@ func (w *writerWithStatus) Header() http.Header {
 }
 
-func MonitorHttp(handler httprouter.Handle, entry *metrics.MetricsEntry) httprouter.Handle {
+func MonitorHttp(handler httprouter.Handle, me *metrics.Metrics, method string) httprouter.Handle {
```
I recommend replacing the string method with constants and using switch statements instead of string compares to offer better performance, which looks to me backtracking on a previous comment. Sigh. So, I think the combined approach here looks good, just replace the string compares with ints for a performance gain.

guscarreon replied:
done

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,198 @@
-	Errors     metrics.Meter
-	BadRequest metrics.Meter
-	Request    metrics.Meter
+// Object to access metric engines
```
Please update comment to Go conventions, something like:
```
Metrics provides access to metric engines.
```

guscarreon replied:
done

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,198 @@
+func CreateMetrics(cfg config.Configuration) *Metrics {
+	engineList := make([]CacheMetrics, 0, 2)
+
+	if cfg.Metrics.Influx.Host != "" {
```
If it's correct that we've validated the metric config by this point, then consider using the .Enabled flag for both influx and prometheus.

guscarreon replied:
done

SyntaxNode commented on metrics/influx/influx.go:
```diff
@@ -0,0 +1,213 @@
+
+var TenSeconds time.Duration = time.Second * 10
+
+//	Object definition
```
Please remove this comment. It is associated with the InfluxMetrics by godocs.

guscarreon replied:
done

SyntaxNode commented on metrics/metricstest/metrics_mock.go:
```diff
@@ -0,0 +1,118 @@
+	"time"
+)
+
+// Define Mock metrics
```
Please remove comment. It's associated with the MockHistograms var.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+	"time"
+)
+
+// Constants and global variables
```
Please remove the "header" comments in this file:

- Constants and global variables
- Object definition
- Init functions
- Helper Init functions
- `CacheMetrics` interface implementation

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+//	Init functions
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	cacheWriteTimeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	requestSizeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1} // TODO: tweak
```
Please resolve this TODO as we're getting close to merging. I think you'd be fine coverage a range up to 1MB, broken up into a max of 10 segments.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+
+//	Init functions
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	cacheWriteTimeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
```
Nitpick: Can be renamed `timeBuckets` for simplicity.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+		Registry: registry,
+		Puts: &PrometheusRequestStatusMetric{
+			Duration: newHistogram(cfg, registry,
+				"puts_current_url_duration",
```
We may choose this time to improve the wording of "current_url" throughout the Prometheus side. We need to leave it alone on the Influx side for compatibility, but here consider using a more obvious term such as "request" like "puts_request_duration" and "puts_request". Leaving the "backend" indicator for outgoing metrics.

guscarreon replied:
done

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,198 @@
+	RecordGetBackendTotal()
+	RecordGetBackendDuration(duration *time.Time)
+	RecordGetBackendError()
+	IncreaseOpenConnections()
```
Consider renaming IncreaseOpenConnections -> RecordConnectionOpen, DecreaseOpenConnections -> RecordConnectionClosed.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+			),
+		},
+		Connections: &PrometheusConnectionMetrics{
+			ConnectionsOpened: newGaugeMetric(cfg, registry,
```
Please replace the connections guage with connections opened and connections closed counters as we have in Prebid Server. The value of the guage can be inferred by Prometheus given the raw data.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus_test.go:
```diff
@@ -0,0 +1,344 @@
+
+					pm.RecordPutDuration(&now)
+				},
+				expDuration:      9.223372036854776e+09,
```
Where does this odd value come from?

guscarreon replied:
That's the value that the `prometheus.Histogram` holds. I'm not sure if the Histogram averages the quanttities before returning the value from the `Histogram.Value` function or if time.Second * 10 gets converted to that number from the beginning. That's the value with which the `assert.Equal` test could pass.

hhhjort replied:
This doesn't seem right. `pm.RecordPutDuration()` is expecting a duration, not a time 10 seconds from now. Shouldn't you be calling `pm.RecordPutDuration(&tenSeconds)` ... Your magic number might be the max possible duration?

guscarreon replied:
`pm.RecordPutDuration` expects a Time pointer. Unfortunately I named it `duration *time.Time` hence the confusion. Do you wnat me to change it to something else? Too `timeLapse *time.Time` maybe?
```
218 func (m *PrometheusMetrics) RecordPutDuration(duration *time.Time) {
219     m.Puts.Duration.Observe(time.Since(*duration).Seconds())
220 }
metrics/prometheus/prometheus.go
```
>... Your magic number might be the max possible duration?

I guess so, I'm honestly not sure.

hhhjort replied:
Ok, now I see. You may want to call the variable `startTime` or something like that. The function being tested looks at the difference between the current time and the time supplied. So in your test you are setting the clock 10 seconds in the future, and then the function tries to determine how much time has past since that future date. It seems that it takes the absolute value, and reports the difference in nanoseconds. Since by the time the function checks the time, it is not quite 10 seconds ahead of the current time, and thus we see "almost 10 seconds in nanoseconds" as the magic number. This is fragile, because if it takes more or less time to get from recording the time until recording the metric, this number will change. Probably should just check it is within some range.

Also should subtract 10 seconds instead of add, so that you are recording a time since some point in the past, rather than "time since some time in the future" since the latter is not well defined.

guscarreon replied:
Substituted `UpdateSince(ts time.Time)` for `Update(d time.Duration)` in metrics/influxdb/influxdb so the interface receives a `time.Duration` parameter instead of `time.Time` object which simplifies testing a lot. More inforation in a slack message I sent to the team's channel.

SyntaxNode commented on metrics/prometheus/prometheus_test.go:
```diff
@@ -0,0 +1,344 @@
+
+	m.RecordExtraTTLSeconds(5)
+	assertHistogram(t, "Assert the extra time to live in seconds was logged", m.ExtraTTL.ExtraTTLSeconds, 2, 5005.00)
+}
```
Please add a gatekeeping test akin to TestMetricCountGatekeeping to ensure accidental high cardinality metrics are not added in the future.

guscarreon replied:
Added. Please let me know if you are ok with this much simpler approach compared to the `prebid-server` version

SyntaxNode commented on metrics/influx/influx.go:
```diff
@@ -0,0 +1,213 @@
+		ExtraTTL:    &InfluxExtraTTL{ExtraTTLSeconds: metrics.GetOrRegisterHistogram("extra_ttl_seconds", r, metrics.NewUniformSample(5000))},
+	}
+
+	metrics.RegisterDebugGCStats(m.Registry)
```
Prometheus is currently missing these. Please check if you can get these all from using `NewGoCollector()`.
https://github.com/prometheus/client_golang/blob/aa9238db679fc02bf22cf4f2c27a980edcb5ada0/prometheus/go_collector.go#L64

guscarreon replied:
I was about to add both `prometheus.NewProcessCollector` and `prometheus.NewGoCollector()` because to the best of my knowledge they are the equivalent of `metrics.CaptureRuntimeMemStats` and `metrics.CaptureDebugGCStats` on the influx side. But at the end I didn't add them because, according to GoDoc site, these two already come in the default registry as described in https://godoc.org/github.com/prometheus/client_golang/prometheus (paragraph shown below). Therefore, I believe CPU usage, and Ram can be collected without the need of explicitly declaring them nor assigning them to the custom `registry` we created in line 63 of `metrics/prometheus/prometheus.go`. Further evidence of this hypothesis is the fact that in `prebid-server` side neither `prometheus.NewProcessCollector` nor `prometheus.NewGoCollector` are ever found in the code base and still we effectively collect CPU and Ram usage for every pod.
> "Also note that the DefaultRegisterer comes registered with a Collector for Go runtime metrics (via NewGoCollector) and a Collector for process metrics (via NewProcessCollector). With a custom registry, you are in control and decide yourself about the Collectors to register."

At the end I didn't add these two at the bottom of `CreatePrometheusMetrics` found in `metrics/prometheus/prometheus.go`. Please let me know if you are ok with this or if you want me to explicitly add them.
```
	// Should be the equivalent of the following influx collectors
	// go metrics.CaptureRuntimeMemStats(m.Registry, flushTime)
	// go metrics.CaptureDebugGCStats(m.Registry, flushTime)
	registry.MustRegister(
		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: cfg.Namespace}),
		prometheus.NewGoCollector(),
	)
```

SyntaxNode replied:
We need to explicitly register the `NewProcessCollector` and the `NewGoCollector` since we're using a custom registry instead of the default registry.

guscarreon replied:
Done

SyntaxNode commented on metrics/core.go:
```diff
@@ -1,119 +1,198 @@
+	RecordGetBackendTotal()
+	RecordGetBackendDuration(duration *time.Time)
+	RecordGetBackendError()
+	IncreaseOpenConnections()
```
Consider renaming IncreaseOpenConnections -> RecordConnectionOpen, DecreaseOpenConnections -> RecordConnectionClosed. These better describe the action to record instead of what the metrics engine needs to do.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,302 @@
+package metrics
```
Do you think it makes sense to 'preload' the values as we do for Prebid Server to avoid the initial setup performance hit

guscarreon replied:
Stole your preload functions from `prebid-server`. But, to my understanding, only `CounterVec`s and `HistogramVec`s need to be preloaded, right? Do I need to do something about non-vector `Histogram` and `Counter` metrics?

SyntaxNode replied:
Yes, that's right. No preload is needed for non-vector `Histogram` and `Counter` metrics.

guscarreon commented on config/config.go:
```diff
@@ -136,37 +136,67 @@ const (
+		log.Fatalf(`Despite being enabled, prometheus metrics came with an empty subsystem value: config.metrics.prometheus.subsystem = \"\".`)
+	}
+	if promMetricsConfig.Enabled {
+		log.Infof("config.metrics.prometheus.namespace: %s", promMetricsConfig.Namespace)
```
done

guscarreon commented on endpoints/decorators/metrics.go:
```diff
@@ -28,9 +28,9 @@ func (w *writerWithStatus) Header() http.Header {
 }
 
-func MonitorHttp(handler httprouter.Handle, entry *metrics.MetricsEntry) httprouter.Handle {
+func MonitorHttp(handler httprouter.Handle, me *metrics.Metrics, method string) httprouter.Handle {
```
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,300 @@
+			),
+		},
+	}
+	promMetrics.ExtraTTL.ExtraTTLSeconds.Observe(5000.00)
```
Was this for debugging purposes?

guscarreon replied:
No, it was there before this PR. Was there on the Influx side so I translated it to prometheus

SyntaxNode replied:
I don't understand it's purpose. Please consider removing it.

SyntaxNode replied:
Please remove this.

SyntaxNode commented on config/config.go:
```diff
@@ -136,37 +136,68 @@ const (
+}
+
+type PrometheusMetrics struct {
+	Port             int    `mapstructure:"port"`
```
Where are you registering the prometheus http handler?

guscarreon replied:
In `server/prometheus.go` and `server/server.go`. The files where there but they needed a little tweak

SyntaxNode commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
+		promMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, prometheus metrics came with an empty port number: config.metrics.prometheus.port = 0`)
+	}
+	if promMetricsConfig.Namespace != "" {
```
This comparison is incorrect. Should be `promMetricsConfig.Namespace == ""`. I'm guessing the logic was reversed in an earlier draft.

guscarreon replied:
corrected

SyntaxNode commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
+		promMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, prometheus metrics came with an empty name space: config.metrics.prometheus.namespace = \"\".`)
+	}
+	if promMetricsConfig.Subsystem != "" {
```
This comparison is incorrect. Should be `promMetricsConfig. Subsystem == ""`. I'm guessing the logic was reversed in an earlier draft.

guscarreon replied:
corrected

SyntaxNode commented on server/server.go:
```diff
@@ -52,7 +55,22 @@ func Listen(cfg config.Configuration, handler http.Handler, metrics *metrics.Con
 	// Once they're finished shutting down (the "done" channel gets pinged for each server),
 	// this funciton can return.
-	wait(stopSignals, done, stopMain, stopAdmin)
+	if cfg.Metrics.Prometheus.Port != 0 {
```
This should only need to check `cfg.Metrics.Enabled` since it will be forced to false by the validator if the port is 0. If the prometheus metrics are disabled this code will currently fail on a nil registry.

guscarreon replied:
Done

SyntaxNode commented on config/config_test.go:
```diff
@@ -116,12 +121,18 @@ backend:
+    namespace: "default-prometheus-namespace"
+    subsystem: "default-prometheus-subsystem"
+    timeout_ms: 100
+    enabled: true
```
Please update config.yaml this way too, except to set the enabled flags to false.

SyntaxNode commented on endpoints/decorators/metrics.go:
```diff
@@ -2,11 +2,40 @@ package decorators
+
+func assignMetricsFunctions(m *metrics.Metrics, method int) *metricsFunctions {
+	metrics := &metricsFunctions{}
+	switch method {
```
IMHO this looks very clean. Love it.

SyntaxNode commented on endpoints/decorators/metrics_test.go:
```diff
@@ -2,69 +2,117 @@ package decorators
 )
 
-func TestSuccessMetrics(t *testing.T) {
+/*
```
Please remove comment.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus_test.go:
```diff
@@ -0,0 +1,356 @@
+	// - This assertion provides a warning for newly added high-cardinality non-adapter specific metrics. The hardcoded limit
+	//   is an arbitrary soft ceiling. Thought should be given as to the value of the new metrics if you find yourself
+	//   needing to increase this number.
+	assert.True(t, generalCardinalityCount <= 52, fmt.Sprintf("General Cardinality. Expected to hace less than 24 metrics, we have: %d \n", generalCardinalityCount))
```
This test is now failing with the addition of the go metrics collectors. There's no need to make this an exact number. Its purpose is to protect against high cardinality. Consider setting to 100 or so as a rough limit.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,320 @@
+	// go metrics.CaptureRuntimeMemStats(m.Registry, flushTime)
+	// go metrics.CaptureDebugGCStats(m.Registry, flushTime)
+	promMetrics.Registry.MustRegister(
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: cfg.Namespace}),
```
For NewProcessCollector, the namespace needs to be manually constructed. It should be cfg.Namespace + _ + cfg.Subsystem. Only include the underscore if neither the namespace or subsystem is null.

guscarreon replied:
done

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,320 @@
+	// go metrics.CaptureDebugGCStats(m.Registry, flushTime)
+	promMetrics.Registry.MustRegister(
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{Namespace: cfg.Namespace}),
+		prometheus.NewGoCollector(),
```
The NewGoCollector does not allow you to provide a namespace. Let's chat about this one in person.

SyntaxNode replied:
We discussed to leave out the GoCollector and follow-up in a future PR since it doesn't allow namespace or subsystem config.

guscarreon commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
+	prometheusEnabled := cfg.Prometheus.Enabled && cfg.Prometheus.validateAndLogMetricsData()
+
+	if !influxEnabled && !prometheusEnabled {
+		log.Fatalf(`No metrics correctly specified in configuration file`)
```
done

guscarreon commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
+}
+
+func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
+	if influxMetricsConfig.Host != "" {
```
done

guscarreon commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
+		influxMetricsConfig.Enabled = false
+		log.Fatalf(`Despite being enabled, influx metrics came with no host info: config.metrics.influx.host = "".`)
+	}
+	if influxMetricsConfig.Database != "" {
```
done

guscarreon commented on config/config.go:
```diff
@@ -136,37 +137,72 @@ const (
 )
 
 type Metrics struct {
-	Type   MetricsType `mapstructure:"type"`
```
This was a bug. Corrected

guscarreon commented:
Addressed issues and ran the server with Prometheus configured to port 9000 to make sure metrics could be displayed
```
|metrics:
|  prometheus:
|    port: 9000
|    namespace: "prebid"
|    subsystem: "cache"
|    timeout_ms: 1000 
|    enabled: true
~ 
config.yaml
```
Then, if we run the server, metrics seem to display correctly:
```
╰─➤  curl http://localhost:9000
# HELP prebid_cache_connection_closed Count the number of closed connections
# TYPE prebid_cache_connection_closed counter
prebid_cache_connection_closed 7
# HELP prebid_cache_connection_error Count the number of connection accept errors or connection close errors
# TYPE prebid_cache_connection_error counter
prebid_cache_connection_error{connection_error="accept"} 0
prebid_cache_connection_error{connection_error="close"} 0
# HELP prebid_cache_connection_opened Count the number of open connections
# TYPE prebid_cache_connection_opened counter
prebid_cache_connection_opened 7
# HELP prebid_cache_extra_ttl_seconds Extra time to live in seconds specified
# TYPE prebid_cache_extra_ttl_seconds histogram
prebid_cache_extra_ttl_seconds_bucket{le="0.001"} 0
prebid_cache_extra_ttl_seconds_bucket{le="0.002"} 0
  .
  .
  .
prebid_cache_puts_request_duration_bucket{le="+Inf"} 2
prebid_cache_puts_request_duration_sum 0.000514076
prebid_cache_puts_request_duration_count 2
```

SyntaxNode commented on config/config.go:
```diff
@@ -136,19 +138,30 @@ const (
+func (cfg *Metrics) checkMetricsEnabled() error {
+	var err error
+	cfg.Influx.Enabled = (cfg.Type == MetricsInflux || cfg.Influx.Enabled)
+	cfg.Influx.Enabled = cfg.Influx.validateAndLogMetricsData()
```
These two Influx lines look a bit strange to me. I think it's because we're using `cfg.Influx.Enabled` as implicit input and an error return channel, in addition to setting it within the method. Consider a flow like this:

```
func (cfg *Metrics) validate() {
  cfg.Influx.Enabled = (cfg.Type == MetricsInflux || cfg.Influx.Enabled)
  if (cfg.Influx.Enabled) {
    cfg.Influx.Enabled = cfg.Influx.validateAndLogMetricsData()
  }

  if (cfg.Prometheus.Enabled) {
    cfg.Influx.Enabled = cfg.Prometheus.validateAndLogMetricsData()
  }
}
```

guscarreon replied:
Done.

SyntaxNode commented on config/config.go:
```diff
@@ -136,19 +138,30 @@ const (
+	cfg.Prometheus.Enabled = cfg.Prometheus.Enabled && cfg.Prometheus.validateAndLogMetricsData()
+
+	if !cfg.Influx.Enabled && !cfg.Prometheus.Enabled && cfg.Type != MetricsNone {
+		err = fmt.Errorf("No metrics correctly specified in configuration file")
```
This should be changed to a log warning. I don't consider it an error to run PBC without any metrics. Since we account for `cfg.Type` in `cfg.Influx.Enabled`, this can just be `!cfg.Influx.Enabled && !cfg.Prometheus.Enabled`.

guscarreon replied:
This logic in line 161 was already in place before this PR and I believe the reason behind it is: if a metrics system other than influx or prometheus was specified, execution stops with an error so we don't expect to get metrics that won't be there. The fact that `cfg.Type != MetricsNone` evaluate to true, opens up the possibility for `cfg.Type` to equal `"Trendalyze"`, `"Amazon Timestream"`, or any other metric system that we don't support.  I prefer the system to use the `cfg.Influx.Enabled` and `cfg.Prometheus.Enabled` flags instead of using the `cfg.Type` field but we decided to let it there for backwards compatibility.
```
155 func (cfg *Metrics) checkMetricsEnabled() error {
156     var err error
157 +--  3 lines: cfg.Influx.Enabled = (cfg.Type == MetricsInflux || cfg.Influx.Enabled)--
160
161     if !cfg.Influx.Enabled && !cfg.Prometheus.Enabled && cfg.Type != MetricsNone {
162         err = fmt.Errorf("No metrics correctly specified in configuration file")
163     }
164     return err
165 }
config/config.go
```
I can still change it to a warning. What do you think?

guscarreon replied:
I believe the new logic illustrates this better. Modified.

SyntaxNode commented on config/config.go:
```diff
@@ -158,15 +171,59 @@ const (
+	Enabled  bool   `mapstructure:"enabled"`
+}
+
+func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
```
Consider renaming to something like `validateAndLog`. It's already associated with the metrics config object.

guscarreon replied:
Done

SyntaxNode commented on config/config.go:
```diff
@@ -158,15 +171,59 @@ const (
+
+func (influxMetricsConfig *InfluxMetrics) validateAndLogMetricsData() bool {
+	if influxMetricsConfig.Host == "" {
+		influxMetricsConfig.Enabled = false
```
Instead of alerting the config directly, it might be better to return a boolean value and let the caller decide what to do with the info.

guscarreon replied:
makes sense, done

SyntaxNode commented on config/config.go:
```diff
@@ -158,15 +171,59 @@ const (
+	Enabled          bool   `mapstructure:"enabled"`
+}
+
+func (promMetricsConfig *PrometheusMetrics) validateAndLogMetricsData() bool {
```
Same feedback as the influx version with the name and changing the struct `Enabled`.

guscarreon replied:
Modified to be more straightforward. Let me know what you think

SyntaxNode commented on metrics/influx/influx.go:
```diff
@@ -0,0 +1,224 @@
+// Export begins sending metrics to the configured database.
+// This method blocks indefinitely, so it should probably be run in a goroutine.
+func (m InfluxMetrics) Export(cfg config.Metrics) {
+	if cfg.Influx.Host != "" {
```
I'm not sure you need this check. I don't believe this method will be called if the host is missing or if influx itself is disabled.

guscarreon replied:
Good catch. Removed

SyntaxNode commented on metrics/metricstest/metrics_mock.go:
```diff
@@ -0,0 +1,132 @@
+	MetricsName string
+}
+
+func (m *MockMetrics) Export(cfg config.Metrics) {
```
Super Nitpick: Consider adding a line break between each method for readability.

guscarreon replied:
This is not exactly in practice throughout the repo so a PR devoted to adding this spaces to the functions in all files would be better. Edited the PR to add a line break in most code I touched though.

SyntaxNode commented on metrics/prometheus/prometheus.go:
```diff
@@ -0,0 +1,296 @@
+
+func CreatePrometheusMetrics(cfg config.PrometheusMetrics) *PrometheusMetrics {
+	timeBuckets := []float64{0.001, 0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
+	requestSizeBuckets := []float64{0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 1}
```
This needs to be updated because the current range is between 0 and 1 bytes. I'll assume the actual size ranges we're monitoring is close to what others hosts are seeing. I recommend:

0, 4kb, 8kb, 16kb, 32kb, 64kb, 128kb, 256kb, 512kb, 1mb

`[]float64{0, 4096, 8192, 16384, 32768, 65536, 131072, 262144, 393216, 524288, 1048576}`
